### DISCUSSION TITLE
Provide a Platform-Agnostic File System API (or Documentation)

### DISCUSSION THREAD

**Discussion by joe-bell** [withastro/roadmap#578] [Proposal]
# Summary

A lot of folks like myself are keen to dive into the world of Edge, but it can be daunting when you're used to using Node internals for accessing raw files (e.g. actual asset buffers, not paths)

It would be wonderful if Astro could provide an API for accessing files that works across Node and Edge environments; making sure that the upgrade path as smooth as possible 

# Background & Motivation

I've been playing around with the "Assets" integration recently (which I know doesn't currently work on Edge due to https://github.com/lovell/sharp/pull/3522#issuecomment-1519857398) and in that process I was trying to `import` the **_actual_** asset  – without much luck

I tried fetching the file in the [same vein as `@vercel/og`](https://vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples#using-a-local-image) but this resulted in an error that "fetch wasn't implemented yet"

Documentation around accessing files in Edge environments is fairly limited (in most frameworks). Having some support to fallback on would also be greatly appreciated

# Goals

Solve the learning curve when moving to Edge-based environments – through documentation, or via a unified file system API

**Comment by lilnasy** [comment]
I can think of an interface but I don't think it is necessary...
### A unified interface
```ts
const file: File | undefined = await Astro.readFile("../../src/assets/image.jpg")
```
...where [file is the web standard object](https://developer.mozilla.org/en-US/docs/Web/API/File), but there's a limitation to how universal this can be, because in serverless environments it will not find anything to return. I'm not sure if this accommodates more use cases than importing.

In your case, you can use the `src` to the original image that `astro:assets` provides. For example, here's how you would let your users open the unprocessed image in a new tab.

```astro
---
import { Image } from "astro:assets"
import myImage from "../assets/my_image.png"
---
<a href={myImage.src} target="_blank">
    <Image src={myImage} alt="A description of my image." />
</a>
```

**Reply by joe-bell** [reply]
> In your case, for example, you can use the src to the original image that astro:assets provides. For example, here's how you would let your users open the unprocessed image in a new tab.

That is indeed currently possible, but my request is specifically around accessing the actual file within the app itself rather than the reference path

**Reply by lilnasy** [reply]
What's your use case?

**Reply by joe-bell** [reply]
1. Providing a dynamic OG endpoint that uses a local image (similar to [this `@vercel/og` use-case](https://vercel.com/docs/concepts/functions/edge-functions/og-image-generation/og-image-examples#using-a-local-image))
2. Creating low-quality image placeholders for local images

Currently getting around this via the Node.js `fs`, however I would like to achieve this in the Edge environment

**Reply by delucis** [reply]
The way I’ve used local images dynamically is to do stuff like glob imports:

```js
const assets = import.meta.glob('../path/to/assets/dir/**/*');
```

You can then find the dynamically passed path in the `assets` object. I believe `import.meta.glob` gets compiled by Vite for SSR targets?

Still has limitations obviously. I guess the main thing is how we mark parts of the file system for inclusion in the SSR bundle. `import` statements and Vite’s `import.meta.glob` are statically analyzable so Astro knows to include anything imported in build output. There’ll always be a line drawn _somewhere_ about what parts of the file system can be accessed, but it would definitely be nice to have simpler APIs for these kinds of use cases.
